include/cpr/threadpool.h

```diff
@@ -0,0 +1,244 @@
+        auto task = std::make_shared<std::packaged_task<RetType()> >(
+            std::bind(std::forward<Fn>(fn), std::forward<Args>(args)...));
+        std::future<RetType> future = task->get_future();
+        {
+            std::lock_guard<std::mutex> locker(task_mutex);
+            tasks.emplace([task]{
+                (*task)();
+            });
+        }
+
+        task_cond.notify_one();
+        return future;
+    }
+
+protected:
+    bool createThread() {
+        if (cur_thread_num >= max_thread_num) return false;
+        std::thread* thread = new std::thread([this] {
+            while (status != STOP) {
+                while (status == PAUSE) {
+                    std::this_thread::yield();
+                }
+
+                Task task;
+                {
+                    std::unique_lock<std::mutex> locker(task_mutex);
+                    task_cond.wait_for(locker, std::chrono::milliseconds(max_idle_time), [this]() {
+                        return status == STOP || !tasks.empty();
+                    });
+                    if (status == STOP) return;
+                    if (tasks.empty()) {
+                        if (cur_thread_num > min_thread_num) {
+                            delThread(std::this_thread::get_id());
+                            return;
+                        }
+                        continue;
+                    }
+                    --idle_thread_num;
+                    task = std::move(tasks.front());
+                    tasks.pop();
+                }
+                if (task) {
+                    task();
+                    ++idle_thread_num;
+                }
+            }
+        });
+        addThread(thread);
+        return true;
+    }
+
+    void addThread(std::thread* thread) {
+        thread_mutex.lock();
+        ++cur_thread_num;
+        ++idle_thread_num;
+        ThreadData data;
+        data.thread = std::shared_ptr<std::thread>(thread);
+        data.id = thread->get_id();
+        data.status = RUNNING;
+        data.start_time = time(nullptr);
+        data.stop_time = 0;
+        threads.emplace_back(data);
+        thread_mutex.unlock();
+    }
+
+    void delThread(std::thread::id id) {
+        time_t now = time(nullptr);
+        thread_mutex.lock();
+        --cur_thread_num;
+        --idle_thread_num;
+        auto iter = threads.begin();
+        while (iter != threads.end()) {
+            if (iter->status == STOP && now > iter->stop_time) {
+                if (iter->thread->joinable()) {
+                    iter->thread->join();
+                    iter = threads.erase(iter);
+                    continue;
+                }
+            } else if (iter->id == id) {
+                iter->status = STOP;
+                iter->stop_time = time(nullptr);
+            }
+            ++iter;
+        }
+        thread_mutex.unlock();
+    }
+
+public:
+    int min_thread_num;
+    int max_thread_num;
+    int max_idle_time;
+
+protected:
```

Comment:
Any reason this can not be:
```suggestion
private:
```